### PR TITLE
HIVE-24693: Convert timestamps to zoned times without string operations

### PR DESCRIPTION
**Pull request by belugabehr**
Replaces #1918

### CODE REVIEW

**Comment by belugabehr** [comment]
I think we have some flaky tests.  Will re-build.

**Comment by belugabehr** [comment]
Before this PR, Hive was not parsing negative dates correctly.  Now it has to because the "parse" code has been removed and it is now just uses the raw values (which are negative numbers).  I changed the formatters to accept negative (and year zero).

I had to update the QTests unfortunately because the 'mask' UDF was assuming that the year 0001 was the first valid year.  The function of the 'mask' method is to transform they year (for example) 2021 -> 0000 in order to "mask" it.  However, since it did not respect/understand/support year '0000', the year is displayed as 0001.  This is a bit confusing from a user standpoint.  Anyway, Hive should be able to support a year 0000, and these code changes require it, so I have updated the qtests to expect a year of value 0000.

https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Types
>The range of values supported for the Date type is 0000-­01-­01 to 9999-­12-­31

**Comment by belugabehr** [comment]
@klcopp @miklosgergely @jcamachor OK.  Finally got this working after many hours of debugging.  Can you please take a look and let me know what you think and if the fundamental changes (unit test changes) are acceptable.

**Comment by klcopp** [comment]
I've seen a couple users come back and ask why 0 shows up as 0001 but they seemed satisfied with the explanation that year 0 doesn't exist. I mean they don't represent all other users but... since it's not a real year I'm not sure we should let users use it.

According to the wiki Hive doesn't support dates/timestamps outside of years 0001–9999. AFAIK currently Hive accepts negative years (though I'm not sure they're displayed/stored correctly?) and auto-converts 0000 to 0001 since 0000 doesn't exist. I think we should decide what we want and change either the wiki or Hive's behavior to not accept pre-0001 and post-9999 dates – don't know how feasible this is though.

@jcamachor  I'd be interested in what you think too!

**Comment by belugabehr** [comment]
@klcopp Where do see about validate date ranges starting at year 1?  I found this on the Hive Wiki:

https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Types

> The range of values supported for the Date type is 0000-­01-­01 to 9999-­12-­31

**Comment by klcopp** [comment]
Touché.
Do users use year 0000 as some sort of null value by the way?

**Comment by belugabehr** [comment]
Honestly, this masking feature seems a bit weird to me.  As a user, I would expect xxxx-xx-xx and not 0000-01-01 but OK.

I don't know, what the correct behavior is, but since it is contrived, I would rather the date parsing be fast, and accurate, than to worry about this masking bit.  Just my opinion.

File: common/src/java/org/apache/hadoop/hive/common/type/TimestampTZUtil.java [hunk 1]
```diff
@@ -127,12 +127,15 @@ public static TimestampTZ parseOrNull(String s, ZoneId defaultTimeZone) {
...
-    return parse(date.toString(), defaultTimeZone);
+    return new TimestampTZ(ZonedDateTime.ofInstant(Instant.ofEpochMilli(date.toEpochMilli()), ZoneOffset.UTC)
+        .withZoneSameLocal(defaultTimeZone));
   }
 
   // Converts Timestamp to TimestampTZ.
   public static TimestampTZ convert(Timestamp ts, ZoneId defaultTimeZone) {
-    return parse(ts.toString(), defaultTimeZone);
+    return new TimestampTZ(
+        ZonedDateTime.ofInstant(Instant.ofEpochSecond(ts.toEpochSecond(), ts.getNanos()), ZoneOffset.UTC)
```
Comment (pgaref) [inline common/src/java/org/apache/hadoop/hive/common/type/TimestampTZUtil.java:137]: use same style one both?

File: common/src/java/org/apache/hadoop/hive/common/type/Date.java [hunk 2]
```diff
@@ -51,7 +51,7 @@
         .appendValue(DAY_OF_MONTH, 1, 2, SignStyle.NORMAL);
     PARSE_FORMATTER = builder.toFormatter().withResolverStyle(ResolverStyle.LENIENT);
     builder = new DateTimeFormatterBuilder();
-    builder.append(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+    builder.append(DateTimeFormatter.ofPattern("uuuu-MM-dd"));
```
Comment (pgaref) [inline common/src/java/org/apache/hadoop/hive/common/type/Date.java:54]: update javadoc above?

File: common/src/java/org/apache/hadoop/hive/common/type/Timestamp.java [hunk 3]
```diff
@@ -72,7 +72,7 @@
     PARSE_FORMATTER = builder.toFormatter().withResolverStyle(ResolverStyle.LENIENT);
     builder = new DateTimeFormatterBuilder();
     // Date and time parts
-    builder.append(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
+    builder.append(DateTimeFormatter.ofPattern("uuuu-MM-dd HH:mm:ss"));
```
Comment (pgaref) [inline common/src/java/org/apache/hadoop/hive/common/type/Timestamp.java:75]: update doc with details?

**Review by pgaref — COMMENTED** [review]
Change makes sense to me thanks @belugabehr -- just thinking if it could have any other side effects

File: common/src/java/org/apache/hadoop/hive/common/type/TimestampTZUtil.java (hunk 1 above)
Comment (belugabehr) [reply common/src/java/org/apache/hadoop/hive/common/type/TimestampTZUtil.java:137]: Hey, what do you mean by same style?

**Comment by belugabehr** [comment]
@pgaref I pushed another commit that simplifies the builder code and adds some more detailed comments.  In generating the comments, I learned something interesting.

```java
private static final DateTimeFormatter PARSE_FORMATTER = new DateTimeFormatterBuilder()
      // Date
      .appendValue(YEAR, 1, 10, SignStyle.NORMAL).appendLiteral('-').appendValue(MONTH_OF_YEAR, 1, 2, SignStyle.NORMAL) ...

private static final DateTimeFormatter PRINT_FORMATTER = new DateTimeFormatterBuilder()
      // Date and Time Parts
      .append(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")) ...
```

When the PARSING code is built, it uses YEAR.  However the FORMATTER code is using "yyyy".  The equivalence is:

`ChornoField.YEAR` = "uuuu"
`ChronoField.YEAR_OF_ERA` = "yyyy"

So, what I ran into in my work on skipping the timestamp parsing is that I stumbled on the fact that Hive is reading YEAR but displaying YEAR_OF_ERA, which are not the same things.  YEAR has negative dates, YEAR_OF_ERA does not usually have negative dates, for example a "negative date" in YEAR_OF_ERA would be +2000 BCE whereas YEAR would be -2000.  So, Hive is kinda whacky and out of sync currently for negative years.

**Comment by belugabehr** [comment]
@klcopp @pvary @ashutosh-bapat ^^^

**Comment by belugabehr** [comment]
Test framework is currently down to validate most recent changes:

```
[2021-02-12T15:48:09.767Z] java.net.ProtocolException: Expected HTTP 101 response but was '500 Internal Server Error'
[2021-02-12T15:48:09.768Z] 	at okhttp3.internal.ws.RealWebSocket.checkResponse(RealWebSocket.java:229)
[2021-02-12T15:48:09.768Z] 	at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:196)
[2021-02-12T15:48:09.768Z] 	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:203)
[2021-02-12T15:48:09.768Z] 	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
[2021-02-12T15:48:09.768Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2021-02-12T15:48:09.768Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2021-02-12T15:48:09.768Z] 	at java.lang.Thread.run(Thread.java:748)
io.fabric8.kubernetes.client.KubernetesClientException: error dialing backend: EOF
```

**Comment by belugabehr** [comment]
```none
[2021-02-12T22:05:12.716Z] java.net.ProtocolException: Expected HTTP 101 response but was '500 Internal Server Error'
[2021-02-12T22:05:12.716Z] 	at okhttp3.internal.ws.RealWebSocket.checkResponse(RealWebSocket.java:229)
[2021-02-12T22:05:12.716Z] 	at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:196)
[2021-02-12T22:05:12.716Z] 	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:203)
[2021-02-12T22:05:12.716Z] 	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
[2021-02-12T22:05:12.716Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2021-02-12T22:05:12.717Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2021-02-12T22:05:12.717Z] 	at java.lang.Thread.run(Thread.java:748)
```
Seems like something wrong with the build.  Will try again.

**Comment by belugabehr** [comment]
` The build of this commit was aborted`

Le sigh

**Review by pgaref — COMMENTED** [review]
Hey @belugabehr changes LGTM and DATE->TS conversion is definitely more efficient.
Regarding masking, I do not consider it an issue as the behaviour can be explained 
 -- @jcamachor did some work there in the past so may also confirm.

**Review by pgaref — APPROVED** [review]

**Pull request merged**